### ISSUE TITLE
Exception handling/logging is incorrect

### ISSUE THREAD

**Issue by ludoch** [GoogleCloudPlatform/appengine-java-vm-runtime#22]
From a customer:
Message: The remote RPC to the application server failed for the call datastore_v3.Get().

com.google.apphosting.api.ApiProxy$RPCFailedException: The remote RPC to the application server failed for the call datastore_v3.Get().
   at com.google.apphosting.vmruntime.VmApiProxyDelegate.runSyncCall(VmApiProxyDelegate.java:17)

It is random, not easily reproducible test case, but they are OK for us to access all possible logs when it happens.
The exact same Java Application (only change is vm=true), so this is a critical issue to fix before moving to GA.

Another app has:

Log/Graph data: com.google.apphosting.api.ApiProxy$RPCFailedException: The remote RPC to the application server failed for the call datastore_v3.BeginTransaction().
at com.google.apphosting.vmruntime.VmApiProxyDelegate.runSyncCall(VmApiProxyDelegate.java:182)
at com.google.apphosting.vmruntime.VmApiProxyDelegate.makeApiCall(VmApiProxyDelegate.java:141)
at com.google.apphosting.vmruntime.VmApiProxyDelegate.access$000(VmApiProxyDelegate.java:47)
at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:360)
at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:336)
at java.util.concurrent.FutureTask.run(FutureTask.java:262)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
at java.lang.Thread.run(Thread.java:745)

Our guess is that datastore has a small background error rate (which is fine). The problem is that we throw the wrong exception "com.google.apphosting.api.ApiProxy$RPCFailedException" instead of one of the errors thrown in V1 (for example com.google.appengine.api.datastore.DatastoreFailureException).

How can we update the Java code to handle these datastore errors better?

**Comment by erugeri** [comment]
FYI, we found a workaround (kind of ugly) for all these heavy read queries made from a backend.

Instead of doing a big findAll that will read all the entities at once, we manually sliced the queries into 1000 entities read steps. We chain them using a filter > last read id value + 1000 limit.

I don't think it's just a bad exception name as in V1 sandbox we very rarely had DatastoreFailureException but in MVM we had many many RPCFailedException like 100 times more.

**Comment by janbartel** [comment]
Was the stacktrace that is shown generated by application code, or by the app engine session manager?  Is the thread that experienced the exception an application thread, or has it been annointed as an app engine server thread?

Jan

**Comment by ludoch** [comment]
@erugeri: would you have a reproducible case? We want to resolve this!

**Comment by erugeri** [comment]
I'm out of office currently. Did you try heavy reads (>100k entities read every 15') ?

**Comment by erugeri** [comment]
We're having a lot of these in our app on production right now. 
This is a **huge** problem.

After 60s (usually, the request takes less than 50ms)
```java
com.google.apphosting.api.ApiProxy$RPCFailedException: The remote RPC to the application server failed for the call taskqueue.BulkAdd().
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.runSyncCall(VmApiProxyDelegate.java:182)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.makeApiCall(VmApiProxyDelegate.java:141)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.access$000(VmApiProxyDelegate.java:47)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:375)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:351)
```

Also, reads and writes to datastore are very very long. A task queue of around 20 000 tasks is usually empty in less than 45 minutes. Today, it took more than 8 hours and we had to raise the max concurrent from 8 to 24.

```java
com.google.apphosting.api.ApiProxy$RPCFailedException: The remote RPC to the application server failed for the call datastore_v3.RunQuery().
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.runSyncCall(VmApiProxyDelegate.java:182)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.makeApiCall(VmApiProxyDelegate.java:141)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.access$000(VmApiProxyDelegate.java:47)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:375)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:351)
```
and
```java
com.google.apphosting.api.ApiProxy$RPCFailedException: The remote RPC to the application server failed for the call datastore_v3.Put().
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.runSyncCall(VmApiProxyDelegate.java:182)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.makeApiCall(VmApiProxyDelegate.java:141)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.access$000(VmApiProxyDelegate.java:47)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:375)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:351)
```

If you need more details or want to look at our app live, don't hesitate to contact me, I'd be happy to give you my phone number.

Please, allow me to sleep tonight .... ;)

**Comment by ludoch** [comment]
what is the app id? not sure it is related to the java runtime but looks
more like an global env issue.

**Comment by erugeri** [comment]
Thanks for your quick reply, I just sent it to you via mail (not sure how safe it is to share it publicly)
Nothing is mentioned here: https://status.cloud.google.com/
Do you know where we could get more info about the issue (is it acknowledged yet? will a fix be released soon? ...). It's quite stressful to see your app being unable to operate, especially with revenue generating requests.
Thanks!

**Comment by step76** [comment]
I have the same problem with my Java APP: the APP works fine on regular instances, but if I switch to managed VM I get ton of similar errors about datastore, taskqueue, memcache, blobstore (to create keys). I've tr
So this is a stopper for us.

**Comment by erugeri** [comment]
Thanks @step76 ! We were feeling very lonely.

FYI, we purchased a support package and we're now in contact with Cloud Platform support team. They unfortunately don't seem to really acknowledge the issue yet :( The number of exceptions exploded (x100) since the 29th September. There were a few errors before but it was manageable thanks to automatic retries.

Hopefully they will be able to find a fix or revert the change they made that day but currently we mostly receive replies saying that Managed VM is in beta and that no SLA is provided :(

**Comment by ludoch** [comment]
Sorry about that...We are looking at it. To be exhaustive, would you mind attaching small reproducible test app?, we would add it in the test env.
Thanks...

**Comment by erugeri** [comment]
I'll try to make a managed vm java app that writes and read continuously on the datastore and add many tasks. Hopefully the bug we'll reproduce.

**Comment by step76** [comment]
I don't know how to create a test case because the errors rises only during peak traffic, otherwise everything is smooth and great.

**Comment by erugeri** [comment]
Here's a very simple app to reproduce the error: https://github.com/erugeri/rpc-exceptions-test

I tried it, and after the #359 refresh (the page auto refresh), there was an error 500: ```com.google.apphosting.api.ApiProxy$RPCFailedException: The remote RPC to the application server failed for the call taskqueue.BulkAdd().```

So the error was on the task queue but you could get it on the writes or the reads.
Don't hesitate if you have any question.

**Comment by ludoch** [comment]
Things are progressing. The bug does not seem to be inside the Jetty process, but in the cloud environment provided by the GCE VM.
Google Folks are working on a fix around the clock now, as it is suddenly more visible than before. Stay tune.

**Comment by erugeri** [comment]
Great news! Thanks for the help, really appreciated!

**Comment by suttiwat** [comment]
Hi :  I also having this problem and posted in the stackoverflow also  -- http://stackoverflow.com/questions/33028819/managed-vm-issue-frequently-found-error-log-about-vmapiproxydelegate-using-dat

I also have question -- Should I add retry-mechanism for RPCFailedException ? Is it useful to add retry? I don't see any documentation from the Google AppEngine document on this.

**Comment by erugeri** [comment]
FYI, the problem is still alive on our app, no changes so far :(

**Comment by suttiwat** [comment]
Yeah me too,  and retrying on RPCFailedException is not 100% guarantee to work. I 've seen a lot of VmApiProxyDelegate  "Read timed out"   on Datastore operations, Memcached operations, Taskqueue operations.

So hope Google engineer fix this soon.

**Comment by ludoch** [comment]
The fix has been applied on the GCE side... Now, I am not sure how long it will hit production (days...hopefully)

**Comment by erugeri** [comment]
Thanks a lot for the update @ludoch

I also received a reply from the support, I really hope you'll be the one who's right : "I will update you once the fix is out. That may take some time, so please expect monthly responses."

Anyway, good to know that a fix was made, thanks again!!

**Comment by suttiwat** [comment]
I'm not sure that the issue has been fixed yet.  But what I noticed right now is for some operations that always failed (cause by this symtom) it's able to run as usual since last Sunday. 

Anyway, some other operations still got the same log of VmApiProxyDelegate.runSyncCall  (Read time out)

**Comment by ludoch** [comment]
The fix is in multiple areas, and the code has been integrated, but some
components are not in prod yet...
On this project, the code change is in
https://github.com/GoogleCloudPlatform/appengine-java-vm-runtime/commit/e78eeef2583dd6ed8fe787ac591f4408a818cbb9
which is giving a better api error when there is an IOException in the
generic comm layer.
Now this IOException should never happen (or with an extreme low rate) in
the future when all the code is in prod.
So yes, the change is not 100% live yet, hopefully next week or the one
after to hit all the prod layers. Maybe more maybe less, I cannot tell:)

Keep an eye on this and let us know if you see issues related to the fix
above, or if you hit different errors (but related).

**Comment by erugeri** [comment]
Hi,
We definitly saw an improvement in the past weeks.
However, this night seemed to be worse than previously and in general the RPC error rate is still higher than 2 months ago. But maybe the fixs aren't fully live yet?
Enjoy Devoxx @ludoch!

**Comment by erugeri** [comment]
It seems exception names have changed :

```java
Message: RCP Failure for API call: datastore_v3 Put
com.google.appengine.api.datastore.DatastoreFailureException: RCP Failure for API call: datastore_v3 Put
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.constructApiException(VmApiProxyDelegate.java:246)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.runSyncCall(VmApiProxyDelegate.java:195)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.makeApiCall(VmApiProxyDelegate.java:154)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.access$000(VmApiProxyDelegate.java:60)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:442)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:418)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
```

```java
Message: RCP Failure for API call: taskqueue BulkAdd
com.google.appengine.api.taskqueue.TransientFailureException: RCP Failure for API call: taskqueue BulkAdd
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.constructApiException(VmApiProxyDelegate.java:219)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.runSyncCall(VmApiProxyDelegate.java:195)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.makeApiCall(VmApiProxyDelegate.java:154)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate.access$000(VmApiProxyDelegate.java:60)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:442)
        at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:418)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
```

What's painful is that we can't catch these exceptions and retry like we did in the past :(
I'm not exactly sure why.
Since yesterday evening, we started to see a lot more :( Please, tell me this is not the final state for MVM.

**Comment by step76** [comment]
I've tried just now and it's totaly unusable for us! After ten minutes a lot of errors on datastore, taskqueue, urlfetch.
We are stuck on classic sandbox!

**Comment by ludoch** [comment]
It is my understanding that not all layers are in prod yet...

**Comment by erugeri** [comment]
Situation is a bit better today.
Hopefully all layers will be live soon. Do you have ETA?
Thanks!

**Comment by suttiwat** [comment]
@erugeri ,  @ludoch 

Today until now (8.00pm Thailand Time, GMT+7) , I've got high exception rate of "DatastoreFailureException" (see stacktrace below),   RCP failure??  or RPC Failure.   

So don't know what happen on the MVM/Compute Engine side or not?  

com.wat_suttiwat.investment.datamanager.CoreDataManager handleDatastoreExceptionLogic: RCP Failure for API call: datastore_v3 Put com.google.appengine.api.datastore.DatastoreFailureException: RCP Failure for API call: datastore_v3 Put at com.google.apphosting.vmruntime.VmApiProxyDelegate.constructApiException(VmApiProxyDelegate.java:246) at com.google.apphosting.vmruntime.VmApiProxyDelegate.runSyncCall(VmApiProxyDelegate.java:195) at com.google.apphosting.vmruntime.VmApiProxyDelegate.makeApiCall(VmApiProxyDelegate.java:154) at com.google.apphosting.vmruntime.VmApiProxyDelegate.access$000(VmApiProxyDelegate.java:60) at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:442) at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:418) at java.util.concurrent.FutureTask.run(FutureTask.java:262) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) at java.lang.Thread.run(Thread.java:745)

**Comment by LusciousPear** [comment]
FYI, I'm encountering similar problems running my test app on localhost. 

    [INFO] com.google.apphosting.api.ApiProxy$RPCFailedException: The remote RPC to the application server failed for the call taskqueue.BulkAdd().
    [INFO] 	at com.google.apphosting.vmruntime.VmApiProxyDelegate.runSyncCall(VmApiProxyDelegate.java:182)
    [INFO] 	at com.google.apphosting.vmruntime.VmApiProxyDelegate.makeApiCall(VmApiProxyDelegate.java:141)
    [INFO] 	at com.google.apphosting.vmruntime.VmApiProxyDelegate.access$000(VmApiProxyDelegate.java:47)
    [INFO] 	at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:371)
    [INFO] 	at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:347)
    [INFO] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    [INFO] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    [INFO] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)

**Comment by LusciousPear** [comment]
Note: restarting and updating to 1.9.28 on localhost seems to have resolved it.

**Comment by suttiwat** [comment]
Latest, I noticed new exception name.   "com.google.apphosting.api.ApiProxy$ArgumentException: An error occurred parsing (locally or remotely) the arguments to DATASTORE_V3.Get()."

com.wat_suttiwat.investment.datamanager.CoreDataManager handleDatastoreExceptionLogic: An error occurred parsing (locally or remotely) the arguments to DATASTORE_V3.Get(). com.google.apphosting.api.ApiProxy$ArgumentException: An error occurred parsing (locally or remotely) the arguments to DATASTORE_V3.Get(). at com.google.apphosting.vmruntime.VmApiProxyDelegate.convertApiResponseRpcErrorToException(VmApiProxyDelegate.java:405) at com.google.apphosting.vmruntime.VmApiProxyDelegate.convertRemoteError(VmApiProxyDelegate.java:349) at com.google.apphosting.vmruntime.VmApiProxyDelegate.runSyncCall(VmApiProxyDelegate.java:187) at com.google.apphosting.vmruntime.VmApiProxyDelegate.makeApiCall(VmApiProxyDelegate.java:154) at com.google.apphosting.vmruntime.VmApiProxyDelegate.access$000(VmApiProxyDelegate.java:60) at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:442) at com.google.apphosting.vmruntime.VmApiProxyDelegate$MakeSyncCall.call(VmApiProxyDelegate.java:418) at java.util.concurrent.FutureTask.run(FutureTask.java:262) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) at java.lang.Thread.run(Thread.java:745)

**Comment by LusciousPear** [comment]
Yes, I've gotten that exact same message.

**Comment by suttiwat** [comment]
As I stay with this problem since early October 2015 until now,  Here below is what I observed from log.

- The problem start when I perform batch-operation such reporting, calculating something. I've variety of batch operation which use only 1 thread and 8 or 16 threads.  I will get the 3 exception patterns below ...

> com.google.appengine.api.datastore.DatastoreFailureException: RCP Failure for API call: datastore_v3 Put.   

> com.google.apphosting.vmruntime.VmApiProxyDelegate runSyncCall: HTTP ApiProxy I/O error for datastore_v3.Put: Read timed out

> com.google.apphosting.api.ApiProxy$ArgumentException: An error occurred parsing (locally or remotely) the arguments to DATASTORE_V3.Get()

It also happened with datastore operations such get, put and other datastore operations.

My gut feeling told me that it might be related with underlying configuration of MVM on the Compute Engine when connect to DataStore Service.   Some configuration about performance-configuration such time-out value,  queue size, whatever performance metrices has been mis-configured.  So it make the datastore operation timeout and raised the exception to upper level.   And sometime I noticed this problem pattern occurred on other network services too such MemCached,  TaskQueue.

I believe there is something wrong the configuration-side. 

This is not happened on the Front-end Instance.  It start to occur this problem on MVM since October 2015 I noticed. before that time MVM work well so good. 

I really like MVM concept but this problem make me feel sick with MVM. Maybe need to take a look at Backend instance instead.

**Comment by LusciousPear** [comment]
Yes, all of that is happening to me as well. Since my app is highly MVM dependent (Jetty8, etc) I don't have another choice.

**Comment by ludoch** [comment]
Some update:
4 Changes were involved for the fix. They got pushed las week, but I just learned that 1 of them was rollbacked and I did not know that.
I am told it is going to be pushed to prod today.
After this is done, the suggestion is that your redeploy your application, or you'll have to wait for a week to see the change. 
I'll ping the thread when the last fix is in prod.

**Comment by erugeri** [comment]
Thanks @ludoch !

**Comment by ludoch** [comment]
The push is done....
Please, try to deploy a new version if you can and let us know!
Potential error rate should now be in the 0.0x% level
A subsequent fix in Cloud Datastore will eventually move it to 0%...

**Comment by erugeri** [comment]
I just deployed both our modules. I'll keep you posted.
Thanks a lot for the news.

**Comment by LusciousPear** [comment]
Is this reflected in the SDK as well?

On Thu, Nov 19, 2015 at 12:05 PM, Etienne Rugeri <notifications@github.com>
wrote:

> I just deployed both our modules. I'll keep you posted.
> Thanks a lot for the news.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/appengine-java-vm-runtime/issues/22#issuecomment-158181405>
> .
>



-- 
Bradford Stephens
Freelance CTO & Startup Exec
22acacia.com
(530) 763-DATA

**Comment by erugeri** [comment]
Our site is down, I don't have more info right now, but there's definitely a HUGE issue in the fix.

**Comment by ludoch** [comment]
bummer. How do you deploy? Via appcfg or gcloud (or gcloud maven plugin) ?

**Comment by ludoch** [comment]
Send me the appid as well (privately if needed)

**Comment by erugeri** [comment]
Situation is back up. More details:

At 21:26 UTC: I receive an downtime alert from Stackdriver and another uptime checker service.
When I go on our website we only get:

Error: Server Error
The service you requested is not available yet.
Please try again in 30 seconds.

I checked on compute engine console: our 4 N2-Standard-2 frontend instances have been rebooted (automatically by the healthcheck? Or for another reason?) But still nothing was working.

I restarted the old instances deployed before and set them 'default', situation seems ok now.

We were on a 46 days uptime run, so that's very rare for us.

We use glcoud maven plugin with the following command: appengine:update

I'l send our appid via email.

**Comment by erugeri** [comment]
We're at the 4th downtime period. Around 10'-15' each time. Suddenly, without any reasons we start to get only 502 responses.
Only way to solve it: redeploy
Does anyone else has the same issue with MVM right now?

**Comment by suttiwat** [comment]
I've tried to redeploy (mvn gcloud:deploy)  and found error during the redeploy below ... 

[INFO] Updating module [tmf-batch]...\
[INFO] Updating module [tmf-batch]...|Deleted [https://www.googleapis.com/compute/v1/projects/xxx-xxx-xxx(censor)/zones/us-central1-f/instances/gae-builder-vm-1].
[INFO] 
[INFO] Updating module [tmf-batch].../
[INFO] Updating module [tmf-batch]...-

[INFO] Updating module [tmf-batch]...\ERROR: (gcloud.preview.app.deploy) Unexpected Error. Version: tmf-batch:1.388690827553976582
[ERROR] Error: gcloud app command with exit code : 1
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------

And the service is not running now (503 Service Error).  I'm redeploying again.  Will let you all know the update soon.

**Comment by suttiwat** [comment]
OK. 2nd retry to re-deploy (mvn gcloud:deploy)  No error that make me shock in first round   :D  

But when I perform 2 type of test.
1st test:   Open URL in browser to access my health check (my custom healthcheck,  not related to MVM health check), it's display 503 Server Error. The service you requested is not available yet

2nd test:  I look into MVM app log.  I've found that my background process running since 9:34AM (GMT+7) to 9:40AM and see   com.google.apphosting.utils.servlet.VmStopFilter doFilter: Running shutdown hook.      

Then I look at shutdown.log   and found 

09:42:37.631  okApp shutdown handlers done.
09:42:37.631  ---------------App was unhealthy, grabbing debug logs----------------

It reboot itself again.  And the same,  my background activity is running (I see from the log of its activity) but I can't access my healthcheck by URL  , browser return 503 Server Error.

**Comment by suttiwat** [comment]
Here below is confirmed behavior I noticed from log.

It reboots again and again (seems reboot every 7 minutes). Don't know the MVM HealthCheck module command this or not. 

So now (11.20AM)  I've updated pom configuration change everything to latest update 1.9.28 and change gcloud-maven-plugin to latest version  2.0.9.87.v20151118  (from 2.0.9.74.v20150814)

Redeploying again ... the result is
1. The 1st deploy always failed, 2nd deploy will be OK.  CONFIRMED.

2. WebInterface on MVM instance always return 503 Server Error  (sometime I saw this strange message -->  Error from HTTP-over-RPC proxy: No response   )   But one thing I observed if I disable MVM HealthCheck in appengineweb.xml ,  WebInterface is return 200 OK as usual (stackdriver cloud monitor can monitor the URL on MVM as usual).   This is CONFIRMED.

3. After run without MVM HealthCheck,  my uptime now is 10 minutes,   No auto-reboot event occurred.   (not see   com.google.apphosting.utils.servlet.VmStopFilter doFilter: Running shutdown hook  in app log anymore)

If it's auto-reboot again, I will post update...

**Comment by erugeri** [comment]
FYI, we're unable to have a stable situation. the module is crashing every 2 hours.
It seems linked to a docker memory issue. When Java reaches 3.9Gb docker kills it. Even on a n1-standard-4 instance (15Gb ram).
Java -Xmx is set to Xmx11660M on a n1-standard-4, however docker seems to have a lower limit.
The error could be found in the syslog file in the instance. We see:

```
 kernel: [12697.633490] [ pid ]   uid  tgid total_vm      rss nr_ptes swapents oom_score_adj name
 kernel: [12697.642974] [ 2294]     0  2294   427732     9093      78        0          -900 docker
 kernel: [12697.652760] [ 3214]     0  3214    16744      617      20        0             0 memcachep
 kernel: [12697.662415] [ 3370]     0  3370    31161     6214      27        0          -900 exe
 kernel: [12697.671650] [ 3380]     0  3380     5013       64      14        0             0 start_nginx.sh
 kernel: [12697.681843] [ 3396]     0  3396    23004      535      43        0             0 nginx
 kernel: [12697.691268] [ 3397] 65534  3397    27162     3371      51        0             0 nginx
 kernel: [12697.700709] [ 3401]     0  3401    12965     2493      29        0             0 gunicorn
 kernel: [12697.710419] [ 3463]     0  3463    60288     9098      53        0             0 gunicorn
 kernel: [12697.720124] [14508]     0 14508     5017      105      14        0             0 jetty_run.sh
 kernel: [12697.730261] [14542]     0 14542  3914288  2679860    5767        0             0 java
 kernel: [12697.739598] Memory cgroup out of memory: Kill process 14542 (java) score 994 or sacrifice child
 kernel: [12697.749866] Killed process 14542 (java) total-vm:15657152kB, anon-rss:10719440kB, file-rss:0kB
```

**Comment by suttiwat** [comment]
OK Update situation from my side since I've stay with it 9-10 hours after new deployment. 

All exceptions that I've ever encountered with such Datastore, TaskQueue, ...  now they are all gone.  It's work very perfect like what MVM does before October 2015.   

For the problem that @erugeri  encounter with, I've no issue like him.  Maybe my memory configuration is smaller than his configuration. 

Anyway,  my new problems above, it's still persisted.  I've to turn-off  MVM Healthcheck to avoid the auto-reboot every 7 minutes (because appengine MVM web server 's not running properly -- return 503 Server Error all time if enable MVM Healthcheck).   This is CONFIRMED again.  :)

**Comment by step76** [comment]
I've deployed this morning, all the errors are gone.
Everything is smooth, no reboots on our side.

**Comment by suttiwat** [comment]
One strange point I observed today,   In my environment, I've configured only 1 instance (n1-standard-2).   When look at the Cloud Console -- AppEngine in sub-menu instance and select MVM module.  It display 1 instance (n1-standard-2) correctly. 

But when I go to Cloud Console -- Compute Engine > Submenu VM Instances ,  I see 2 instances.  One is for MVM my app (zone: us-central1-c).  The other one is in zone us-central1-f.   

And when I query with gcloud compute instances list 

> gcloud compute instances list    it list 2 instances 

> gae-tmf--abcde-1-2rg6 us-central1-c n1-standard-2             10.240.0.2  104.197.180.XX RUNNING
> gae-tmf--abcde-1-q5ap us-central1-f n1-standard-1             10.240.0.3  104.197.176.YY RUNNING

So I don't know where the additional instance (suffix:  q5ap) comes from.  And it will charge me or not?  I never see MVM behavior like this before.  Do you have any ideas?


And, lastly,  now (12:45AM, Thailand Time, GMT+7),   I can't re-deploy (mvn gcloud:deploy).  It shows 

[INFO] ERROR: Error Response: [500] Internal error encountered.
[INFO] ERROR: (gcloud.preview.app.deploy) Could not retrieve the default Google Cloud Storage bucket for [censor-app-id]. Please try again or use the [bucket] argument.

**Comment by suttiwat** [comment]
FYI,  update as of Saturday morning 9.00AM (GMT+7) .  Seems there are some update on the google side last night, 

- I can deploy in first time as usual. No need to 2nd retry. 
- The problem about "Could not retrieve the default Google Cloud Storage bucket " gone. 
- No auto-reboot (every 7 minutes) if turn-on MVM HealthCheck. 

Here is  last remaining  the observed mis-behaviors of MVM
1. I've to disable MVM Healthcheck in <enable-health-check>false</enable-health-check>    in order to make Servlet Target on MVM return 200 (OK) | 502 (Bad Gateway with message  "Error from HTTP-over-RPC proxy: No response")        If not, it will always return 503 Server Error all time.
2. When perform "gcloud compute instances list"  ,  I will get the list that has extra instance.  This is not belong to my app.  I'm trying to delete it but can't , it auto-restart many time.     My valid MVM instance is up on zone  us-central1-c.    But the alien instance is up on zone  us-central1-f.    Can't stop, delete it. It's automatically restart.    I'm worry about billing extra-charge  of this alien instance. 

That's the updated from my side. All other things, no datastore, taskqueue exceptions like in previous any more.

**Comment by suttiwat** [comment]
FYI,  Updated as of Nov 21,  9:25PM (GMT+7 time)

- Alien instance gone. I've not do anything from my side.  Just see it stopped state , so I delete instance.  It's not re-born anymore.  the problem has been solved by itself or by Google engineer in 12 hours ago. 

- Servlet on MVM right now always return 200 OK ,  not mixed with 502 (Bad Gateway) anymore -- after alien instance gone.   BUT,  servlet on MVM will return 200 only  I disable MVM healthcheck   (<enable-health-check>false</enable-health-check>)      But if I set it true to turn-on MVM healthcheck,   servlet on MVM will always 100% return 503 Server Error all time.   ** This is final last remaining issue for me **
